File: CSharpCodeBase/utils/hump/timeoutlist.lua
local Class = require 'utils.hump.class'

local TimeoutList = Class({init = function (self, globaltimeout)
      self._items = {}
      self._times = {}
      self._globaltimeout = globaltimeout
end})

function TimeoutList:Exists(item)
  local idx = Tables.Find(self._items, item)
  return idx > - 1
end

function TimeoutList:Add(item, timeout)
  timeout = timeout or self._globaltimeout
  local idx = Tables.Find(self._items, item)
  if idx > -1 then
    self._times[idx] = os.clock() + timeout
  else
    self._items[#self._items + 1] = item
    self._times[#self._times + 1] = os.clock() + timeout
  end
end

function TimeoutList:Remove(item)
  local idx = Tables.Find(self._items, item)
  if idx > -1 then
    table.remove(self._items, idx)
    table.remove(self._times, idx)
  end
end

function TimeoutList:SetGlobalTimeout(timeout)
  self._globaltimeout = timeout
end

function TimeoutList:IsEmpty()
  return #self._items == 0
end

function TimeoutList:Update()
  if self._globaltimeout == -1 then return end
  for k, v in pairs(self._times) do
    if os.clock() > v then
      table.remove(self._items, k)
      table.remove(self._times, k)
      break
    end
  end
end

function TimeoutList:GetData()
  return self._items
end

return TimeoutList

File: CSharpCodeBase/utils/hump/timeoutlist.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class timeoutlist {
 public void init (self, globaltimeout);
       this._items = {}
       this._times = {}
       this._globaltimeout = globaltimeout;
 }})
 public void Exists(item){
   var idx = Tables.Find(this._items, item);
   return idx > - 1;
 }
 public void Add(item, timeout){
   timeout = timeout  ||  this._globaltimeout;
   var idx = Tables.Find(this._items, item);
   if(idx > -1  ){ 
     this._times[idx] = os.clock() + timeout;
   }else{
     this._items[#this._items + 1] = item;
     this._times[#this._times + 1] = os.clock() + timeout;
   }
 }
 public void Remove(item){
   var idx = Tables.Find(this._items, item);
   if(idx > -1  ){ 
     table.remove(this._items, idx);
     table.remove(this._times, idx);
   }
 }
 public void SetGlobalTimeout(timeout){
   this._globaltimeout = timeout;
 }
 public void IsEmpty(){
   return #this._items == 0;
 }
 public void Update(){
   if(this._globaltimeout == -1  ){  return }
   for(k, v in pairs(this._times) ){
     if(os.clock() > v  ){ 
       table.remove(this._items, k);
       table.remove(this._times, k);
       break;
     }
   }
 }
 public void GetData(){
   return this._items;
 }
 }}